File: src/main/scala/com/vc/hard/L588.scala
package com.vc.hard

import scala.collection.mutable
case class Dir(dirs: mutable.HashMap[String, Dir], files: mutable.HashMap[String, String])

class FileSystem() {

  val root = Dir(new mutable.HashMap[String, Dir](), new mutable.HashMap[String, String]())

  def ls(path: String): List[String] = {
    val list = new mutable.ListBuffer[String]()
    var current = root
    val pSplit = path.split("/")
    (1 until pSplit.length).foreach(i => {
      val d = pSplit(i)
      if(current.dirs.contains(d)) {
        current = current.dirs(d)
      }
      else if(current.files.contains(d)) {
        list += d
        return list.toList
      }
      else return list.toList
    })
    current.dirs.foreach(x => list += x._1)
    current.files.foreach(x => list += x._1)
    list.sorted.toList
  }

  def mkdir(path: String) {
    val pSplit = path.split("/")
    var current = root
    (1 until pSplit.length).foreach(i => {
      val d = pSplit(i)
      if(current.dirs.contains(d)) current = current.dirs(d)
      else {
        val dir = new Dir(
          new mutable.HashMap[String, Dir](),
          new mutable.HashMap[String, String]()
        )
        current.dirs.put(d, dir)
        current = current.dirs(d)
      }
    })
  }

  def addContentToFile(filePath: String, content: String) {
    val pSplit = filePath.split("/")
    val file = pSplit(pSplit.length - 1)
    var current = root
    (1 until pSplit.length - 1).foreach(i => {
      val d = pSplit(i)
      if(current.dirs.contains(d)) current = current.dirs(d)
    })
    current.files.put(file, current.files.getOrElse(file, "") + content)
  }

  def readContentFromFile(filePath: String): String = {
    val pSplit = filePath.split("/")
    val file = pSplit(pSplit.length - 1)
    var current = root
    (0 until pSplit.length - 1).foreach(i => {
      val d = pSplit(i)
      if(current.dirs.contains(d)) current = current.dirs(d)
    })
    if(current.files.contains(file)) current.files(file) else ""
  }

}

/**
  * Your FileSystem object will be instantiated and called as such:
  * var obj = new FileSystem()
  * var param_1 = obj.ls(path)
  * obj.mkdir(path)
  * obj.addContentToFile(filePath,content)
  * var param_4 = obj.readContentFromFile(filePath)
  */

File: src/main/java/com/vc/hard/L588.java
package com.vc.hard;

import java.util.*;

class FileSystem {
    static class Dir {
        HashMap<String, Dir> dirs;
        HashMap<String, String> files;

        Dir(HashMap<String, Dir> dirs, HashMap<String, String> files) {
            this.dirs = dirs;
            this.files = files;
        }

        @Override
        public String toString() {
            return "dirs: "+dirs+" files: "+files;
        }
    }

    private Dir root;
    public FileSystem() {
        root = new Dir(new HashMap<String, Dir>(), new HashMap<String, String>());
    }

    public List<String> ls(String path) {
        List<String> res = new ArrayList<String>();

        String[] pathSplits = path.split("/");
        Dir current = root;
        for(int i = 1; i < pathSplits.length; i++) {
            if(current.dirs.containsKey(pathSplits[i])) current = current.dirs.get(pathSplits[i]);
            else if(current.files.containsKey(pathSplits[i])) {
                res.add(pathSplits[i]);
                return res;
            }
            else {
                return res;
            }
        }

        for(Map.Entry<String, String> files: current.files.entrySet()) {
            res.add(files.getKey());
        }

        for(Map.Entry<String, Dir> dir: current.dirs.entrySet()) {
            res.add(dir.getKey());
        }

        Collections.sort(res, new Comparator<String>() {
            public int compare(String f1, String f2) {
                return f1.compareTo(f2);
            }
        });
        return res;
    }

    public void mkdir(String path) {
        String[] pathSplits = path.split("/");
        Dir current = root;
        for(int i = 1; i < pathSplits.length; i++) {
            if(!current.dirs.containsKey(pathSplits[i])) {
                current.dirs.put(pathSplits[i],
                        new Dir(new HashMap<String, Dir>(), new HashMap<String, String>()));
            }
            current = current.dirs.get(pathSplits[i]);
        }
    }

    public void addContentToFile(String filePath, String content) {
        String[] pathSplits = filePath.split("/");
        Dir current = root;
        for(int i = 1; i < pathSplits.length - 1; i++) {
            current = current.dirs.get(pathSplits[i]);
        }

        String fileName = pathSplits[pathSplits.length - 1];
        current.files.put(fileName, current.files.getOrDefault(fileName, "") + content);
    }

    public String readContentFromFile(String filePath) {
        String[] pathSplits = filePath.split("/");
        Dir current = root;
        for(int i = 1; i < pathSplits.length - 1; i++) {
            current = current.dirs.get(pathSplits[i]);
        }
        String fileName = pathSplits[pathSplits.length - 1];
        return current.files.getOrDefault(fileName, "");
    }
}

/**
 * Your FileSystem object will be instantiated and called as such:
 * FileSystem obj = new FileSystem();
 * List<String> param_1 = obj.ls(path);
 * obj.mkdir(path);
 * obj.addContentToFile(filePath,content);
 * String param_4 = obj.readContentFromFile(filePath);
 */
